File: Scaling/genFluidShape.py
import json
import sys
import copy 
import os

# ミキサーの水平，垂直方向の幅の最大値(最小2)
hvMAX = 3
# 液滴数の最大値
fluidNumMax = 10
# 生成したデータを保存するディレクトリのパス
path = './FluidShapeLib'

def isInsideBoundary(i,j):
    if 0<=i and i< hvMAX and 0<=j and j<hvMAX:
        return True 
    else :
        return False 

def main():
    ### 液滴1個は，2個の場合以降の元となるので，最初に単独で生成
    json_data = {}
    for i in range(hvMAX):
        for j in range(hvMAX):
            # 2x2ミキサーが最小探索領域
            SearchSpaceSize = str(max(i+1,j+1,2))
            if SearchSpaceSize not in json_data:
                json_data[SearchSpaceSize] = {}
            stridx = str(len(json_data[SearchSpaceSize]))
            coords = []
            coord = {"y":i,"x":j}
            coords.append(coord)
            json_data[SearchSpaceSize][stridx] = coords

    try:
        os.mkdir(path)
        print("Folder %s created!" % path)
    except FileExistsError:
        print("Folder %s already exists" % path)

    with open("FluidShapeLib/1fluid.json", "w") as f:
        json.dump(json_data, f,  ensure_ascii=False, indent=4,  separators=(',', ': '))
    
    dy = [0,-1,0,1]
    dx = [1,0,-1,0]
    
    # n個の液滴によって構成される形を元に，n+1個の液滴から構成される液滴の形を探索する．
    for fluidnum in range(2,fluidNumMax+1):
        SetOfNewShape = set()
        former = str(fluidnum-1)+"fluid.json"
        new = str(fluidnum)+"fluid.json"
       
        with open(os.path.join(path,former), "r") as f:
            FormerFluidShapes = json.load(f)
            for formerSearchSpace in range(2,hvMAX+1):
                strfss = str(formerSearchSpace)
                if strfss in FormerFluidShapes:
                    for fluididx in range(len(FormerFluidShapes[strfss])):
                        strfluididx = str(fluididx)
                        FluidShape = FormerFluidShapes[strfss][strfluididx]
                        shape = []
                        # shapeの形の情報を冗長なjson形式から圧縮して扱いやすくした（簡易的な座標形式）配列
                        # 液滴はいかなるx,y断面においても，整数の長さの線分とならなくてはならない（飛び地になったらダメ）
                        lowerY,upperY= [-100 for i in range(hvMAX)],[-100 for i in range(hvMAX)]
                        lowerX,upperX = [-100 for i in range(hvMAX)],[-100 for i in range(hvMAX)]
                        for coord in FluidShape:
                            y,x = coord["y"],coord["x"]
                            if lowerX[x]==-1 and upperX[x]==-1:
                                lowerX[x] = y 
                                upperX[x] = y 
                            else: 
                                lowerX[x] = min(y,lowerX[x])
                                upperX[x] = max(y,upperX[x])
                            if lowerY[y]==-1 and upperY[y]==-1:
                                lowerY[y] = x
                                upperY[y] = x
                            else: 
                                lowerY[y] = min(x,lowerY[y])
                                upperY[y] = max(x,upperY[y])
                            shape.append(tuple((y,x)))

                        for coordidx in range(len(shape)):
                            y,x = shape[coordidx]
                            for way in range(4):
                                ny,nx = y+dy[way],x+dx[way]
                                ncoord = (ny,nx)
                                if isInsideBoundary(ny,nx) and ncoord not in shape:
                                    nshape = copy.deepcopy(shape)
                                    nshape.append(ncoord)
                                    nshape.sort()
                                    # セットで扱えないのでnshapeをリストからタプルに詰め替える
                                    TupNShape = tuple(nshape)
                                    # (ny,nx)をshapeに加えたnshapeが有効な形（飛地のない形）かどうかを確認
                                    # ~FirstTime = その行/列に座標を置かれるのが初めて
                                    # ~NextToFormer = その座標の上下/左右の座標がすでに使われている
                                    # xf,xn,yf,ynとすると，((not xf) and (not xn)) or ((not yf) and (not yn)) がTrueになる場合を防ぎたい
                                    # また，アルゴリズム的にないはずだが，(not xf) and (not yf)がTrueになっていないことも確認
                                    nxFirstTime = lowerX[nx]==-100 and upperX[nx]==-100
                                    nxNextToFormer = abs(ny-lowerX[nx])<=1 or abs(ny-upperX[nx])<=1
                                    nyFirstTime = lowerY[ny]==-100 and upperY[ny]==-100
                                    nyNextToFormer = abs(nx-lowerY[ny])<=1 or abs(nx-upperY[ny])<=1
                                    if TupNShape not in SetOfNewShape \
                                        and (((not nxFirstTime) or (not nyFirstTime)) and (nxFirstTime or nxNextToFormer) and (nyFirstTime or nyNextToFormer)) :
                                        SetOfNewShape.add(TupNShape)

        with open(os.path.join(path,new), "w") as nf:
            json_data = {}
            for nshape in SetOfNewShape: 
                SearchSpace = -1
                OutputFormedData = []
                for ncoord in nshape: 
                    y,x = ncoord 
                    SearchSpace = max(SearchSpace,y+1,x+1)
                    res = {"y":y,"x":x}
                    OutputFormedData.append(res)
                # isInsideBoundaryでチェックしているが念のため再度チェック
                if 2<= SearchSpace and SearchSpace <= hvMAX:
                    strss = str(SearchSpace)
                    idx = 0
                    if strss not in json_data: 
                        json_data[strss] = {}
                    else: 
                        idx = len(json_data[strss])
                    json_data[strss][str(idx)] = OutputFormedData 
                else :
                    print("探索領域を超えたサイズの液滴です",nshape,SearchSpace,hvMAX,file=sys.stderr)
            json.dump(json_data, nf,  ensure_ascii=False, indent=4,  separators=(',', ': '))
   
if __name__ == "__main__":
    main()

File: Scaling/genFluidShape.cpp
#include "json.hpp"
#include <sys/stat.h>
#include <algorithm>
#include <fstream>
#include <ios>
#include <iostream>
#include <iterator>
#include <map>
#include <set>
#include <sstream>
#include <string.h>
#include <string>
#include <typeinfo>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
#include <chrono>

using json = nlohmann::json;
using namespace std;

const int fluidNumMax = 10; 
const int hvMAX = 2;
const string path = "./FluidShapeLib/";

bool isInsideBoundary(int y,int x){
    if (0<=y && y< hvMAX && 0<=x && x<hvMAX)
        return true;
    else 
        return false;
}

bool hasKey(string key, json object) {
  auto itr_check = object.find(key);
  if (itr_check != object.end())
    return true;
  else
    return false;
}

bool SameArray(json array, json cmp_array) {
  if (array.size() == cmp_array.size()) {
    for (int i = 0; i < array.size(); i++) {
      if (array[i] != cmp_array[i])
        return false;
    }
    return true;
  }
  return false;
}


bool ArrayContainsValue(json value, json Array) {
  for (auto v : Array) {
    if (v == value) {
      return true;
    }
  }
  return false;
}

// genFluidShape.pyのアルゴリズムをC++に書き直した．
// genFluidShape.pyの方が詳しくコメントを書いている.参照されたし！！
int main() {
    // 液滴1個のデータを作成
    json json_data; 
    for (int i=0; i<hvMAX; i++){
        for(int j=0; j<hvMAX; j++){
            string SearchSpaceSize = to_string(max(max(i+1,j+1),2)); 
            if(!json_data.contains(SearchSpaceSize))
                json_data[SearchSpaceSize] = json::object();
            string stridx = to_string(json_data[SearchSpaceSize].size()); 
            vector<json> coords; 
            map<string,int> res{{"y",i},{"x",j}};
            json coord(res); 
            coords.push_back(coord); 
            json_data[SearchSpaceSize][stridx] = coords; 
        }
    }
    if(mkdir(path.c_str(), 0777)==-1)
        cout<<path<<"を作成しようとしましたが，もうすでに同名のディレクトリが存在しました（特に問題ない場合は無視してください）.\n";
    string WriteFileName = path+"1fluid.json"; 
    ofstream writing_file;
    writing_file.open(WriteFileName,ios::out); 
    writing_file << json_data <<endl; 
    writing_file.close();

    vector<int> dy{0,-1,0,1};  
    vector<int> dx{-1,0,1,0};
    // n個の液滴によって構成される形を元に，n+1個の液滴から構成される液滴の形を探索する．
    for (int fluidnum=2;fluidnum<=fluidNumMax;fluidnum++ ){
        string FormerFilePath = path+to_string(fluidnum-1)+"fluid.json"; 
        string NewFilePath = path+to_string(fluidnum)+"fluid.json"; 

        ifstream reading1(FormerFilePath, ios::in);
        json FormerFluidShapes; 
        reading1 >> FormerFluidShapes;

        set<vector<pair<int,int>> > SetOfNewShape; 
        //set< vector<int> > SetOfNewShape; 
        for (int formerSearchSpace=2;formerSearchSpace<=hvMAX;formerSearchSpace++){
            string strfss = to_string(formerSearchSpace); 
            if (hasKey(strfss,FormerFluidShapes)){
                for(int fluididx=0; fluididx<size(FormerFluidShapes[strfss]);fluididx++){
                    string strfluididx = to_string(fluididx); 
                    vector<json> FluidShape = FormerFluidShapes[strfss][strfluididx]; 
                    // より扱いやすいデータ型に変換
                    vector<pair<int,int>> shape ; 
                    vector<int> lowerY(hvMAX,-100),upperY(hvMAX,-100),lowerX(hvMAX,-100),upperX(hvMAX,-100); 
                    for(json coord:FluidShape){
                        int y=coord["y"],x=coord["x"]; 
                        if (lowerX[x]==-100 && upperX[x]==-100)
                            lowerX[x] = y,upperX[x] = y; 
                        else 
                            lowerX[x] = min(y,lowerX[x]),upperX[x] = max(y,upperX[x]); 
                        if (lowerY[y]==-100 && upperY[y]==-100)
                            lowerY[y] = x,upperY[y] = x; 
                        else
                            lowerY[y] = min(x,lowerY[y]),upperY[y] = max(x,upperY[y]); 
                        pair<int,int> p{y,x}; 
                        shape.push_back(p); 
                    }

                    for(int coordidx=0; coordidx<size(shape); coordidx++){
                        int y=get<0>(shape[coordidx]),x=get<1>(shape[coordidx]); 
                    for(int way=0; way<4; way++){
                        int ny=y+dy[way],nx=x+dx[way]; 
                        pair<int,int> ncoord{ny,nx}; 
                        if(isInsideBoundary(ny,nx) && count(shape.begin(),shape.end(),ncoord)==0){
                            auto nshape = shape; 
                            nshape.push_back(ncoord); 
                            sort(nshape.begin(),nshape.end()); 
                            bool nxFirstTime = (lowerX[nx]==-100&&upperX[nx]==-100); 
                            bool nxNextToFormer = (abs(ny-lowerX[nx])<=1 || abs(ny-upperX[nx])<=1); 
                            bool nyFirstTime = (lowerY[ny]==-100 || upperY[ny]==-100); 
                            bool nyNextToFormer = (abs(nx-lowerY[ny])<=1 || abs(nx-upperY[ny])<=1); 
                            if (!SetOfNewShape.contains(nshape)
                                && ((!nxFirstTime || !nyFirstTime) && (nxFirstTime || nxNextToFormer) && (nyFirstTime || nyNextToFormer)) ){
                                SetOfNewShape.emplace(nshape); 
                            }
                        }
                    }
                    }
                }
            }
        }
        ofstream new_file;
        new_file.open(NewFilePath,ios::out); 
        json new_data; 
        for (auto nshape : SetOfNewShape){
            int SearchSpace = -1; 
            vector<json> OutputFormedData ; 
            for(auto ncoord : nshape){
                int y=get<0>(ncoord),x=get<1>(ncoord); 
                SearchSpace = max(SearchSpace,max(y+1,x+1));
                map<string,int> res{{"y",y},{"x",x}}; 
                json coord(res); 
                OutputFormedData.push_back(coord); 
            }
            if(2<=SearchSpace && SearchSpace<=hvMAX){
                string strss = to_string(SearchSpace); 
                int idx = 0; 
                if(!hasKey(strss,new_data))
                    new_data[strss] = json::object(); 
                else  
                    idx = size(new_data[strss]);
                new_data[strss][to_string(idx)] = OutputFormedData; 
            }
            else  
                cerr<<"探索領域を超えたサイズの液滴です"<<SearchSpace<<hvMAX<<endl; 
        }
        new_file << new_data <<endl; 
        new_file.close();
    }
    cout<<"データの生成完了"<<endl; 
    return 0;
}
